### PR TITLE
[Al Amin Step-10]  Sort your task list by creation time

### PR DESCRIPTION
**Pull request by rr-mdalauddin-alamin**
### Step 10: Sort the task list in order of creation date and time (Optional)

- Currently, they are sorted in order of ID, but let's sort them in descending order of creation date and time.
- Let's write in the system spec that the sorting is working well

### CODE REVIEW

**Comment by ksak19** [comment]
📝 I faced some issues when I tried to run this app on my local. Here is what I did to solve them:
* `myapp/docker-compose.yml`
  * add `platform: linux/amd64` under `db`
  * add `platform: linux/arm64` under `chrome`
  * change host port of `api` from `3001` to `3030` <- `3001` is already used for daily development
* Enter api container
  * `bundle exec rails c`
  * `User.create!(first_name: 'test', last_name: 'test', email: 'test@example.com', password: 'hoge', is_admin: true, username: 'test')`
  * cuz user must exist to add task

File: myapp/app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,38 @@
...
+      Rails.logger.info "Rendering 404 with exception: #{e.message}" if e
+
+      render 'errors/404', status: :not_found
+    end
+
+    def _render_500(e = nil)
+      Rails.logger.error "Rendering 500 with exception: #{e.message}" if e
+
+      render 'errors/500', status: :internal_server_error
+    end
```
Comment (ksak19) [inline myapp/app/controllers/application_controller.rb]:
```suggestion
  private
  
  def _render_404(e = nil)
    Rails.logger.info "Rendering 404 with exception: #{e.message}" if e

    render 'errors/404', status: :not_found
  end

  def _render_500(e = nil)
    Rails.logger.error "Rendering 500 with exception: #{e.message}" if e

    render 'errors/500', status: :internal_server_error
  end
```
In Rakuma, indent level of `private` and methods below it should be the same

Comment (ksak19) [reply myapp/app/controllers/application_controller.rb]: Seems like there is no description about [this](https://github.com/Fablic/training/blob/m-alamin/steps_jp.md#:~:text=Gemfile%20%E3%81%A7%20RuboCop%EF%BC%88Ruby%E3%81%AE%E9%9D%99%E7%9A%84%E3%82%B3%E3%83%BC%E3%83%89%E8%A7%A3%E6%9E%90%E3%83%84%E3%83%BC%E3%83%AB%EF%BC%89%20%E3%81%8B%20fablicop%E3%82%92%E3%82%A4%E3%83%B3%E3%82%B9%E3%83%88%E3%83%BC%E3%83%AB%E3%81%97%E3%81%BE%E3%81%97%E3%82%87%E3%81%86%E3%80%82%E8%A8%AD%E5%AE%9A%E3%82%84%E3%82%B3%E3%83%9E%E3%83%B3%E3%83%89%E5%AE%9F%E8%A1%8C%E6%96%B9%E6%B3%95%E3%81%AF%E5%90%84README%E3%81%8B%E3%82%89%E7%A2%BA%E8%AA%8D%E3%81%97%E3%81%BE%E3%81%97%E3%82%87%E3%81%86%E3%80%82) in English ver README. You should install `fablicop` gem to follow linting rules in Rakuma 😄

File: myapp/app/views/errors/404.html.erb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+<!-- public/404.html -->
+<!DOCTYPE html>
+<html>
+<head>
+  <title><%= t('errors.404_not_found') %></title>
+</head>
+<body>
+<h1><%= t('errors.404_not_found') %></h1>
+</body>
+</html>
```
Comment (ksak19) [inline myapp/app/views/errors/404.html.erb]: you should insert final new line at the end of file

File: myapp/config/routes.rb [hunk 3]
```diff
@@ -3,4 +3,8 @@
 
   resources :task
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+
+  # Custom Error pages handling routes
+  get '*not_found' => 'application#routing_error'
```
Comment (ksak19) [inline myapp/config/routes.rb]: didn't know this 👍

File: myapp/app/controllers/task_controller.rb
Comment (ksak19) [inline myapp/app/controllers/task_controller.rb:1]: file name should be `tasks_controller.rb` (pluralized) based on Rails convention 😅 
Found useful gist: https://gist.github.com/iangreenleaf/b206d09c587e8fc6399e
> Model class names use CamelCase. These are singular, and will map automatically to the plural database table name.
> Controller class names use CamelCase and have Controller as a suffix. The Controller suffix is always singular. The name of the resource is usually plural.

File: myapp/app/controllers/task_controller.rb [hunk 4]
```diff
@@ -1,6 +1,12 @@
 class TaskController < ApplicationController
   def index
-    @tasks = Task.all
+    if params[:sort] == 'latest'
+      @tasks = Task.all.order(created_at: :desc)
```
Comment (ksak19) [inline myapp/app/controllers/task_controller.rb]:
```suggestion
      @tasks = Task.order(created_at: :desc)
```
`all` is unnecessary, I guess?

File: myapp/spec/system/task_sorting_func_spec.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
+require 'rails_helper'
+
+RSpec.describe 'Task Sorting', type: :system do
+  before do
+    # Create a user because we have foreign key
+    create(:user)
+  end
+
+  it 'check sort functionality by items as latest first & oldest first' do
```
Comment (ksak19) [inline myapp/spec/system/task_sorting_func_spec.rb]: I prefer these test cases are written separately
```ruby
it 'check sort functionality by items as latest first' do
  ...
  expect(task_titles_only).to eq(['Test Task latest one', 'Test Task created on second', 'Test Task first one'])
end

it 'check sort functionality by items as oldest first' do
  ...
  expect(task_titles_only).to eq(['Test Task first one', 'Test Task created on second', 'Test Task latest one'])
end
```

Comment (ksak19) [reply myapp/spec/system/task_sorting_func_spec.rb]: Basically we should avoid writing multiple `expect` in one `it` block. Sometimes we use multiple `expect` but test code look like this in that case:
```ruby
it 'expects multiple things' do
 do_something_like_clicking_link
 expect(foo).to eq 'bar'
 expect(bar).to eq 'baz'
end
# write multiple expectation at the end of the block
```

**Review by ksak19 — COMMENTED** [review]
I've learned new things through this review! Left some comments. Thank you

File: myapp/app/controllers/task_controller.rb
Comment (rr-mdalauddin-alamin) [reply myapp/app/controllers/task_controller.rb:1]: Thank you so much

**Comment by mt-ytaher** [comment]
> 📝 I faced some issues when I tried to run this app on my local. Here is what I did to solve them:
> 
> * `myapp/docker-compose.yml`
>   
>   * add `platform: linux/amd64` under `db`
>   * add `platform: linux/arm64` under `chrome`
>   * change host port of `api` from `3001` to `3030` <- `3001` is already used for daily development
> * Enter api container
>   
>   * `bundle exec rails c`
>   * `User.create!(first_name: 'test', last_name: 'test', email: 'test@example.com', password: 'hoge', is_admin: true, username: 'test')`
>   * cuz user must exist to add task

These are really good points. I had to do all these steps as well. so maybe @alaminshakeel  san should add it to the readme for the next reviewers. Also modify the `docker-compose.yml`

**Comment by rr-mdalauddin-alamin** [comment]
> > 📝 I faced some issues when I tried to run this app on my local. Here is what I did to solve them:
> > 
> > * `myapp/docker-compose.yml`
> >   
> >   * add `platform: linux/amd64` under `db`
> >   * add `platform: linux/arm64` under `chrome`
> >   * change host port of `api` from `3001` to `3030` <- `3001` is already used for daily development
> > * Enter api container
> >   
> >   * `bundle exec rails c`
> >   * `User.create!(first_name: 'test', last_name: 'test', email: 'test@example.com', password: 'hoge', is_admin: true, username: 'test')`
> >   * cuz user must exist to add task
> 
> These are really good points. I had to do all these steps as well. so maybe @alaminshakeel san should add it to the readme for the next reviewers. Also modify the `docker-compose.yml`

@taheryasir 
Yeah right, we talked about it. Going to add a `db:seed` and describe in Readme.md. Here we are using `volume` option to keep the data. so we think it would be better describe in Readme.md to run few commands for the first time. Isn't it will be enough? do you have any other thought? 

Thanks

File: myapp/app/controllers/application_controller.rb (hunk 1 above)
Comment (rr-mdalauddin-alamin) [reply myapp/app/controllers/application_controller.rb]: fixed

File: myapp/app/views/errors/404.html.erb (hunk 2 above)
Comment (rr-mdalauddin-alamin) [reply myapp/app/views/errors/404.html.erb]: fixed

File: myapp/app/controllers/task_controller.rb
Comment (rr-mdalauddin-alamin) [reply myapp/app/controllers/task_controller.rb:1]: done!

**Comment by mt-ytaher** [comment]
@alaminshakeel Sounds good. As long as reviewer or anyone can create the environment without any issues or debugging by themselves is fine.

File: myapp/spec/system/task_sorting_func_spec.rb (hunk 5 above)
Comment (rr-mdalauddin-alamin) [reply myapp/spec/system/task_sorting_func_spec.rb]: separated

Comment (rr-mdalauddin-alamin) [reply myapp/spec/system/task_sorting_func_spec.rb]: > Basically we should avoid writing multiple `expect` in one `it` block. Sometimes we use multiple `expect` but test code look like this in that case:
> 
> ```ruby
> it 'expects multiple things' do
>  do_something_like_clicking_link
>  expect(foo).to eq 'bar'
>  expect(bar).to eq 'baz'
> end
> # write multiple expectation at the end of the block
> ```

For the `task_spec.rb` i left as it was but for the `task_sorting_func_spec.rb` I made the separation. 
so please check once again

Thanks

File: myapp/app/controllers/tasks_controller.rb [hunk 6]
```diff
@@ -2,7 +2,10 @@
 
 class TasksController < ApplicationController # rubocop:todo Style/Documentation
   def index
-    @tasks = Task.all
+    @tasks = Task.order(created_at: :asc)
+    return unless params[:sort] == 'latest'
+
+    @tasks = Task.all.order(created_at: :desc)
```
Comment (ksak19) [inline myapp/app/controllers/tasks_controller.rb]: `all` 😄

**Review by ksak19 — APPROVED** [review]
LGTM 👍

File: myapp/app/controllers/tasks_controller.rb (hunk 6 above)
Comment (rr-mdalauddin-alamin) [reply myapp/app/controllers/tasks_controller.rb]: すみません🙇

**Pull request merged**